EDU-182 Add migration script for regenerating cdnResources

adrianaluchian opened the pull request:
Closes https://educandu.atlassian.net/browse/EDU-182

adrianaluchian commented on migrations/manual/educandu-2021-12-14-01-migrate-cdn-resources-once-more.js:
```diff
@@ -0,0 +1,113 @@
+    case 'audio':
+      return [...acc, ...processAudioSection(section)];
+    case 'markdown':
+      return [...acc, ...processMarkdownSection(section)];
```
This is the same migration script content as in `educandu-2021-11-18-02-migrate-cdn-resoures.js`, except for the 'markdown' case which is new.

ahelmberger commented on migrations/manual/educandu-2021-12-14-01-migrate-cdn-resources-once-more.js:
```diff
@@ -0,0 +1,113 @@
+    return content;
+  }
+  const regexp = /(?<=\]\(cdn:\/\/)(.*?)(?=\))/g;
+  const matches = content.match(regexp);
```
:+1:  Yes, also in the test!

adrianaluchian replied:
I was just testing you!
Corrected, thanks!

laurentiu-ilici replied:
I'm glad we passed teacher 😆

ahelmberger commented on migrations/manual/educandu-2021-12-14-01-migrate-cdn-resources-once-more.js:
```diff
@@ -0,0 +1,113 @@
+  if (!content) {
+    return content;
+  }
+  const regexp = /(?<=\]\(cdn:\/\/)(.*?)(?=\))/g;
```
:heart: I see what you did there, wow! From now on I will ask you when I got stuck with my lookbehinds and lookaheads!

ahelmberger commented on migrations/manual/educandu-2021-12-14-01-migrate-cdn-resources-once-more.js:
```diff
@@ -0,0 +1,113 @@
+  }
+}
+
+export default class Educandu_2021_12_14_01_migrate_cdn_resources_once_more {
```
In my opinion that might be a little bit misleading as the actual reason we have to run this again is the hard-delete bug.

laurentiu-ilici replied:
Okay

ahelmberger commented:
It would be cool if we could print to the console all revisions that had changes (and also the changes themselves). There shouldn't be really that many we have to update, so in case we get hundreds we will know that something went wrong. What do you think?

adrianaluchian commented:
> It would be cool if we could print to the console all revisions that had changes (and also the changes themselves). There shouldn't be really that many we have to update, so in case we get hundreds we will know that something went wrong. What do you think?

Sure, we can print, but what do you mean by "There shouldn't be really that many we have to update" ? 
We're basically regenerating all of them.

laurentiu-ilici commented on migrations/manual/educandu-2021-12-14-01-migrate-cdn-resources-once-more.spec.js:
```diff
@@ -0,0 +1,110 @@
+  sourceUrl: 'media/video-content-1.mov'
+};
+const videoContent2 = {
+  sourceType: 'external',
```
There's also the type "youtube" just for info. I think it would be overkill to add it to the tests, but maybe we could take the chance to modify the script to not populate them in the cdn resources when you have an URL marked as "internal" but it has the value of https://youtube/etwas. I think we have a couple of documents that have this problem and it was causing some import issues (at least on local, the real S3 doesn't seem to care).

adrianaluchian replied:
We could do that, sure, but that would not solve the problem unless we also don't allow adding external URLs when 'internal' is selected. So new such cases will be introduced.
We should make a ticket for that, fix the code and add a dedicated migrations script to correct such cases.

adrianaluchian replied:
https://educandu.atlassian.net/browse/EDU-187

adrianaluchian commented on migrations/manual/educandu-2021-12-14-01-migrate-cdn-resources-once-more.js:
```diff
@@ -0,0 +1,113 @@
+  }
+  const regexp = /(?<=\]\(cdn:\/\/)(.*?)(?=\))/g;
+  const matches = content.match(regexp);
+  return matches ? [...new Set(matches)] : [];
```
Removed

adrianaluchian commented:
> > It would be cool if we could print to the console all revisions that had changes (and also the changes themselves). There shouldn't be really that many we have to update, so in case we get hundreds we will know that something went wrong. What do you think?
> 
> Sure, we can print, but what do you mean by "There shouldn't be really that many we have to update" ? We're basically regenerating all of them.

I've updated the script actually (since you mentioned) to only save changes when there are some.
Also added some logging. The problem now is though that I cannot see the logged messages when running `gulp migrate`

laurentiu-ilici reviewed (approved):
Reapprove after the intersection changes
